### PR TITLE
[YUNIKORN-1041] Fix binding of dynamic volumes to pod

### PR DESCRIPTION
**Pull request by craigcondit**
### What is this PR for?
Fix binding of dynamic volumes to pods. This is broken after YUNIKORN-872. Also add additional error logging.

### What type of PR is it?
* [x] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1041

### How should this be tested?
Manual testing for now, e2e tests to be added to YUNIKORN-1039.

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

Error occurring under 0.12.1:

error-0-12-1

Error gone after fix:

no-error-fix

### CODE REVIEW

File: pkg/cache/context.go
Comment (craigcondit) [inline pkg/cache/context.go:414]: This is the actual fix (ensure these slices are empty rather than nil).

File: pkg/cache/context.go
Comment (craigcondit) [inline pkg/cache/context.go:421]: This is the actual fix -- don't allow the slices within volumes to be nil.

File: pkg/cache/context.go
Comment (craigcondit) [inline pkg/cache/context.go:382]: This is based on upstream K8s; if we have unbound immediate claims, we can't satisfy this.

File: pkg/cache/context.go
Comment (craigcondit) [inline pkg/cache/context.go:413]: This was added based on upstream K8s. If any reasons are returned from FindPodVolumes, there are conflicts and we must fail.

Comment (wilfred-s) [reply pkg/cache/context.go:413]: looks good

Comment (yangwwei) [reply pkg/cache/context.go:413]: Good catch.

```
volumes, reasons, err := ctx.apiProvider.GetAPIs().VolumeBinder.FindPodVolumes(assumedPod, boundClaims, claimsToBind, node)
```
Seems like there will be a case where `err==nil` but `len(reasons) > 0`

> // It returns an error when something went wrong or a list of reasons why the node is
	// (currently) not usable for the pod.

File: pkg/cache/context.go
Comment (yangwwei) [reply pkg/cache/context.go:421]: why does this fix the problem? 
I did not see anywhere to complain NIL value

**Review by yangwwei — CHANGES_REQUESTED** [review]

**Comment by yangwwei** [comment]
Do you plan to add any UT or e2e test for this?

File: pkg/cache/context.go
Comment (wilfred-s) [reply pkg/cache/context.go:421]: This prevents the early return in the K8s code. The K8s API code expects an empty array and never a nil for any success case. Even in the K8s test code a nil seems to be replaced with an empty array for success cases if I read the code correctly.

File: pkg/cache/context.go
Comment (wilfred-s) [inline pkg/cache/context.go:369]: This same call is made in `AssumePod` line 457 why do we not need the change there?

File: pkg/cache/context.go
Comment (wilfred-s) [reply pkg/cache/context.go:421]: The nil checks for the `volumes.StaticBindings` and `volumes.DynamicProvisions` have been in the code since k8s 1.12. The real cause is a flow on effect of the change in the `FindPodVolumes` return value as part of a cleanup commit.  That was added in k8s 1.19 and later: 0 length slices are replaced with nil values:
https://gitlab.cncf.ci/kubernetes/kubernetes/commit/2cdc63aeaab603c61b48e95c044f1e4bb2f32548

We need tests for this code path.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
`bindPodVolumes` from the context is called from an event in the task. That task event should have dispatched a failure which I assumed should have failed the whole task which should have triggered a retry etc. 

If there was no failure there is an error handling issue. These changes do not change this at all. Does that mean that volume bind failures are not handled at all?
Having a warning event does not equate to a failure does it?

We should have a test for this

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:421]: > We need tests for this code path.

I don't disagree. However, currently, the entirety of volume binding is disabled in test mode. Mocking out the proper interactions is going to be a pretty major undertaking.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:413]: Exactly.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:380]: I agree, we probably need to test for this earlier. I'll work on trying to get this working earlier as well. Your suggestion as to where makes sense -- I will investigate that.

**Comment by craigcondit** [comment]
> Do you plan to add any UT or e2e test for this?

Yes, e2e tests will be added in YUNIKORN-1039. Those will take time to develop and this issue is a blocker for 0.12.2 so I don't think we should hold up the fix for the e2e tests to be developed.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:369]: We probably do. I missed that one.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:380]: Looking at this a bit more, I don't believe we need to worry about this. The existing VolumeBinder plugin gets called as part of pod / node evaluation, so this logic (which comes from PreFilter) should already fire prior to us attempting to bind the node.

Comment (craigcondit) [reply pkg/cache/context.go:380]: We should still check it during binding to ensure that the status has not changed though.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:369]: actually, i don't think we do. This logic will have already run via the PreFilter() / Filter() calls on the upstream VolumeBinder plugin (we call these as part of node fitness checks).

File: pkg/cache/context.go
Comment (yangwwei) [reply pkg/cache/context.go:380]: I see, thanks. Looks like this is already covered in [predict manager](https://github.com/apache/incubator-yunikorn-k8shim/blob/799a42ed9b247815fd344f70e0df27d6b83f90d2/pkg/plugin/predicates/predicate_manager.go#L86), that's good.

**Review by yangwwei — APPROVED** [review]
Looks good from my side. Thanks for fixing this!

**Comment by craigcondit** [comment]
> `bindPodVolumes` from the context is called from an event in the task. That task event should have dispatched a failure which I assumed should have failed the whole task which should have triggered a retry etc.
> 
> If there was no failure there is an error handling issue. These changes do not change this at all. Does that mean that volume bind failures are not handled at all? Having a warning event does not equate to a failure does it?

`task.postTaskAllocated()` dispatches a `FailTaskEvent` if `err != nil` is returned from `context.bindPodVolumes()`, so I think we're good there.

File: pkg/cache/context.go
Comment (craigcondit) [reply pkg/cache/context.go:421]: Given the major surgery required to get more of context.go tested, I'd like to move forward with this PR as-is to unblock 0.12.2 and add e2e tests in YUNIKORN-1039 targeting 1.0.0. @wilfred-s does that sound ok with you?

**Comment by craigcondit** [comment]
Moving forward with this to unblock 0.12.2 release. @wilfred-s, if you have additional concerns, let's address in a follow-up patch for 1.0.0.

File: pkg/cache/context.go
Comment (wilfred-s) [reply pkg/cache/context.go:421]: yes that is good

**Pull request closed without merging**
